File: canopy/plugin/csharp/src/CanopyPlugin/contract.cs
using System;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Google.Protobuf;
using Google.Protobuf.Reflection;
using Google.Protobuf.WellKnownTypes;
using Types;

namespace CanopyPlugin
{
    // ContractConfig: the configuration of the contract
    public static class ContractConfig
    {
        public const string Name = "csharp_plugin_contract";
        public const int Id = 1;
        public const int Version = 1;
        public static readonly string[] SupportedTransactions = { "send" };
        public static readonly string[] TransactionTypeUrls = 
        { 
            "type.googleapis.com/types.MessageSend"
        };
        public static readonly string[] EventTypeUrls = Array.Empty<string>();
        // Include google/protobuf/any.proto first as it's a dependency of event.proto and tx.proto
        public static readonly ByteString[] FileDescriptorProtos =
        {
            ByteString.CopyFrom(Any.Descriptor.File.ToProto().ToByteArray()),
            ByteString.CopyFrom(AccountReflection.Descriptor.ToProto().ToByteArray()),
            ByteString.CopyFrom(EventReflection.Descriptor.ToProto().ToByteArray()),
            ByteString.CopyFrom(PluginReflection.Descriptor.ToProto().ToByteArray()),
            ByteString.CopyFrom(TxReflection.Descriptor.ToProto().ToByteArray()),
        };
    }

    // Contract defines the smart contract that implements the extended logic of the nested chain
    public class Contract
    {
        private static readonly Random Random = new();

        public Config Config { get; }
        public PluginFSMConfig? FsmConfig { get; }
        public Plugin Plugin { get; }
        public ulong FsmId { get; }

        public Contract(Config config, Plugin plugin, ulong fsmId, PluginFSMConfig? fsmConfig = null)
        {
            Config = config;
            Plugin = plugin;
            FsmId = fsmId;
            FsmConfig = fsmConfig;
        }

        // Genesis implements logic to import a json file to create the state at height 0
        public PluginGenesisResponse Genesis(PluginGenesisRequest request)
        {
            return new PluginGenesisResponse();
        }

        // BeginBlock is code that is executed at the start of applying the block
        public PluginBeginResponse BeginBlock(PluginBeginRequest request)
        {
            return new PluginBeginResponse();
        }

        // CheckTx is code that is executed to statelessly validate a transaction
        public async Task<PluginCheckResponse> CheckTxAsync(PluginCheckRequest request)
        {
            // validate fee
            var resp = await Plugin.StateReadAsync(this, new PluginStateReadRequest
            {
                Keys = { new PluginKeyRead { QueryId = (ulong)Random.NextInt64(), Key = ByteString.CopyFrom(KeyForFeeParams()) } }
            });

            if (resp.Error != null)
            {
                return new PluginCheckResponse { Error = resp.Error };
            }

            // convert bytes into fee parameters
            var minFees = Unmarshal<FeeParams>(resp.Results[0].Entries[0].Value.ToByteArray());
            if (minFees == null)
            {
                return new PluginCheckResponse { Error = ErrUnmarshal("fee params") };
            }

            // check for the minimum fee
            if (request.Tx.Fee < minFees.SendFee)
            {
                return new PluginCheckResponse { Error = ErrTxFeeBelowStateLimit() };
            }

            // handle the message based on type
            var typeUrl = request.Tx.Msg.TypeUrl;
            
            if (typeUrl.EndsWith("/types.MessageSend"))
            {
                var msg = new MessageSend();
                msg.MergeFrom(request.Tx.Msg.Value);
                return CheckMessageSend(msg);
            }
            else
            {
                return new PluginCheckResponse { Error = ErrInvalidMessageCast() };
            }
        }

        // DeliverTx is code that is executed to apply a transaction
        public async Task<PluginDeliverResponse> DeliverTxAsync(PluginDeliverRequest request)
        {
            // handle the message based on type
            var typeUrl = request.Tx.Msg.TypeUrl;
            
            if (typeUrl.EndsWith("/types.MessageSend"))
            {
                var msg = new MessageSend();
                msg.MergeFrom(request.Tx.Msg.Value);
                return await DeliverMessageSendAsync(msg, request.Tx.Fee);
            }
            else
            {
                return new PluginDeliverResponse { Error = ErrInvalidMessageCast() };
            }
        }

        // EndBlock is code that is executed at the end of applying a block
        public PluginEndResponse EndBlock(PluginEndRequest request)
        {
            return new PluginEndResponse();
        }

        // CheckMessageSend statelessly validates a 'send' message
        private PluginCheckResponse CheckMessageSend(MessageSend msg)
        {
            // check sender address
            if (msg.FromAddress.Length != 20)
            {
                return new PluginCheckResponse { Error = ErrInvalidAddress() };
            }

            // check recipient address
            if (msg.ToAddress.Length != 20)
            {
                return new PluginCheckResponse { Error = ErrInvalidAddress() };
            }

            // check amount
            if (msg.Amount == 0)
            {
                return new PluginCheckResponse { Error = ErrInvalidAmount() };
            }

            // return the authorized signers
            return new PluginCheckResponse
            {
                Recipient = msg.ToAddress,
                AuthorizedSigners = { msg.FromAddress }
            };
        }

        // DeliverMessageSend handles a 'send' message
        private async Task<PluginDeliverResponse> DeliverMessageSendAsync(MessageSend msg, ulong fee)
        {
            var fromQueryId = (ulong)Random.NextInt64();
            var toQueryId = (ulong)Random.NextInt64();
            var feeQueryId = (ulong)Random.NextInt64();

            // calculate the from key and to key
            var fromKey = KeyForAccount(msg.FromAddress.ToByteArray());
            var toKey = KeyForAccount(msg.ToAddress.ToByteArray());
            var feePoolKey = KeyForFeePool((ulong)Config.ChainId);

            // get the from and to account
            var response = await Plugin.StateReadAsync(this, new PluginStateReadRequest
            {
                Keys =
                {
                    new PluginKeyRead { QueryId = feeQueryId, Key = ByteString.CopyFrom(feePoolKey) },
                    new PluginKeyRead { QueryId = fromQueryId, Key = ByteString.CopyFrom(fromKey) },
                    new PluginKeyRead { QueryId = toQueryId, Key = ByteString.CopyFrom(toKey) }
                }
            });

            // check for internal error
            if (response.Error != null)
            {
                return new PluginDeliverResponse { Error = response.Error };
            }

            // get the bytes from response
            byte[]? fromBytes = null, toBytes = null, feePoolBytes = null;
            foreach (var result in response.Results)
            {
                if (result.QueryId == fromQueryId)
                    fromBytes = result.Entries.FirstOrDefault()?.Value?.ToByteArray();
                else if (result.QueryId == toQueryId)
                    toBytes = result.Entries.FirstOrDefault()?.Value?.ToByteArray();
                else if (result.QueryId == feeQueryId)
                    feePoolBytes = result.Entries.FirstOrDefault()?.Value?.ToByteArray();
            }

            // convert the bytes to account structures
            var from = new Account();
            var to = new Account();
            var feePool = new Pool();

            if (fromBytes != null && fromBytes.Length > 0)
                from.MergeFrom(fromBytes);
            if (toBytes != null && toBytes.Length > 0)
                to.MergeFrom(toBytes);
            if (feePoolBytes != null && feePoolBytes.Length > 0)
                feePool.MergeFrom(feePoolBytes);

            // add fee to 'amount to deduct'
            var amountToDeduct = msg.Amount + fee;

            // if the account amount is less than the amount to subtract; return insufficient funds
            if (from.Amount < amountToDeduct)
            {
                return new PluginDeliverResponse { Error = ErrInsufficientFunds() };
            }

            // for self-transfer, use same account data
            var isSelfTransfer = fromKey.SequenceEqual(toKey);
            if (isSelfTransfer)
            {
                to = from;
            }

            // subtract from sender
            from.Amount -= amountToDeduct;
            // add the fee to the 'fee pool'
            feePool.Amount += fee;
            // add to recipient
            to.Amount += msg.Amount;

            // execute writes to the database
            var writeRequest = new PluginStateWriteRequest();

            // add fee pool update
            writeRequest.Sets.Add(new PluginSetOp
            {
                Key = ByteString.CopyFrom(feePoolKey),
                Value = ByteString.CopyFrom(from.Amount == 0 ? to.ToByteArray() : feePool.ToByteArray())
            });

            // fix: always write fee pool correctly
            writeRequest.Sets.Clear();
            writeRequest.Sets.Add(new PluginSetOp
            {
                Key = ByteString.CopyFrom(feePoolKey),
                Value = ByteString.CopyFrom(feePool.ToByteArray())
            });

            // if the from account is drained - delete the from account
            if (from.Amount == 0)
            {
                writeRequest.Deletes.Add(new PluginDeleteOp { Key = ByteString.CopyFrom(fromKey) });
            }
            else
            {
                writeRequest.Sets.Add(new PluginSetOp
                {
                    Key = ByteString.CopyFrom(fromKey),
                    Value = ByteString.CopyFrom(from.ToByteArray())
                });
            }

            // write to account (skip if self-transfer since we already handled it)
            if (!isSelfTransfer)
            {
                writeRequest.Sets.Add(new PluginSetOp
                {
                    Key = ByteString.CopyFrom(toKey),
                    Value = ByteString.CopyFrom(to.ToByteArray())
                });
            }

            var writeResp = await Plugin.StateWriteAsync(this, writeRequest);
            return new PluginDeliverResponse { Error = writeResp.Error };
        }

        // State key prefixes
        private static readonly byte[] AccountPrefix = { 0x01 };
        private static readonly byte[] PoolPrefix = { 0x02 };
        private static readonly byte[] ParamsPrefix = { 0x07 };

        // KeyForAccount returns the state database key for an account
        public static byte[] KeyForAccount(byte[] addr)
        {
            return JoinLenPrefix(AccountPrefix, addr);
        }

        // KeyForFeeParams returns the state database key for fee parameters
        public static byte[] KeyForFeeParams()
        {
            return JoinLenPrefix(ParamsPrefix, Encoding.UTF8.GetBytes("/f/"));
        }

        // KeyForFeePool returns the state database key for the fee pool
        public static byte[] KeyForFeePool(ulong chainId)
        {
            return JoinLenPrefix(PoolPrefix, FormatUInt64(chainId));
        }

        // JoinLenPrefix appends the items together separated by a single byte to represent the length
        public static byte[] JoinLenPrefix(params byte[][] items)
        {
            var result = new System.Collections.Generic.List<byte>();
            foreach (var item in items)
            {
                if (item == null || item.Length == 0)
                    continue;
                result.Add((byte)item.Length);
                result.AddRange(item);
            }
            return result.ToArray();
        }

        // FormatUInt64 converts a ulong to big-endian bytes
        public static byte[] FormatUInt64(ulong value)
        {
            var bytes = BitConverter.GetBytes(value);
            if (BitConverter.IsLittleEndian)
                Array.Reverse(bytes);
            return bytes;
        }

        // Marshal serializes a proto.Message into a byte slice
        public static byte[] Marshal(IMessage message)
        {
            return message.ToByteArray();
        }

        // Unmarshal deserializes a byte slice into a proto.Message
        public static T? Unmarshal<T>(byte[] data) where T : IMessage<T>, new()
        {
            if (data == null || data.Length == 0)
                return default;
            var parser = new MessageParser<T>(() => new T());
            return parser.ParseFrom(data);
        }

        // Error factory methods - matching Go implementation
        private const string DefaultModule = "plugin";

        public static PluginError ErrPluginTimeout() =>
            new() { Code = 1, Module = DefaultModule, Msg = "a plugin timeout occurred" };

        public static PluginError ErrMarshal(string err) =>
            new() { Code = 2, Module = DefaultModule, Msg = $"marshal() failed with err: {err}" };

        public static PluginError ErrUnmarshal(string err) =>
            new() { Code = 3, Module = DefaultModule, Msg = $"unmarshal() failed with err: {err}" };

        public static PluginError ErrFailedPluginRead(string err) =>
            new() { Code = 4, Module = DefaultModule, Msg = $"a plugin read failed with err: {err}" };

        public static PluginError ErrFailedPluginWrite(string err) =>
            new() { Code = 5, Module = DefaultModule, Msg = $"a plugin write failed with err: {err}" };

        public static PluginError ErrInvalidPluginRespId() =>
            new() { Code = 6, Module = DefaultModule, Msg = "plugin response id is invalid" };

        public static PluginError ErrUnexpectedFSMToPlugin(string type) =>
            new() { Code = 7, Module = DefaultModule, Msg = $"unexpected FSM to plugin: {type}" };

        public static PluginError ErrInvalidFSMToPluginMessage(string type) =>
            new() { Code = 8, Module = DefaultModule, Msg = $"invalid FSM to plugin: {type}" };

        public static PluginError ErrInsufficientFunds() =>
            new() { Code = 9, Module = DefaultModule, Msg = "insufficient funds" };

        public static PluginError ErrFromAny(string err) =>
            new() { Code = 10, Module = DefaultModule, Msg = $"fromAny() failed with err: {err}" };

        public static PluginError ErrInvalidMessageCast() =>
            new() { Code = 11, Module = DefaultModule, Msg = "the message cast failed" };

        public static PluginError ErrInvalidAddress() =>
            new() { Code = 12, Module = DefaultModule, Msg = "address is invalid" };

        public static PluginError ErrInvalidAmount() =>
            new() { Code = 13, Module = DefaultModule, Msg = "amount is invalid" };

        public static PluginError ErrTxFeeBelowStateLimit() =>
            new() { Code = 14, Module = DefaultModule, Msg = "tx.fee is below state limit" };
    }
}

File: canopy/plugin/kotlin/src/main/kotlin/com/canopy/plugin/Contract.kt
package com.canopy.plugin

import com.google.protobuf.Any
import com.google.protobuf.AnyProto
import com.google.protobuf.ByteString
import mu.KotlinLogging
import types.AccountOuterClass
import types.AccountOuterClass.Account
import types.AccountOuterClass.Pool
import types.EventOuterClass
import types.Plugin
import types.Plugin.*
import types.Tx
import types.Tx.FeeParams
import types.Tx.MessageSend
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.random.Random

private val logger = KotlinLogging.logger {}

/**
 * Contract configuration matching Go's ContractConfig
 */
object ContractConfig {
    const val NAME = "kotlin_plugin_contract"
    const val ID = 1L
    const val VERSION = 1L
    val SUPPORTED_TRANSACTIONS = listOf("send")
    val TRANSACTION_TYPE_URLS = listOf(
        "type.googleapis.com/types.MessageSend"
    )
    val EVENT_TYPE_URLS = emptyList<String>()
    val FILE_DESCRIPTOR_PROTOS = listOf(
        // Include google/protobuf/any.proto first as it's a dependency of event.proto and tx.proto
        AnyProto.getDescriptor().toProto().toByteString(),
        AccountOuterClass.getDescriptor().toProto().toByteString(),
        EventOuterClass.getDescriptor().toProto().toByteString(),
        Plugin.getDescriptor().toProto().toByteString(),
        Tx.getDescriptor().toProto().toByteString(),
    )

    fun toPluginConfig(): PluginConfig = PluginConfig.newBuilder()
        .setName(NAME)
        .setId(ID)
        .setVersion(VERSION)
        .addAllSupportedTransactions(SUPPORTED_TRANSACTIONS)
        .addAllTransactionTypeUrls(TRANSACTION_TYPE_URLS)
        .addAllEventTypeUrls(EVENT_TYPE_URLS)
        .addAllFileDescriptorProtos(FILE_DESCRIPTOR_PROTOS)
        .build()
}

// Key prefixes matching Go implementation
private val ACCOUNT_PREFIX = byteArrayOf(1)
private val POOL_PREFIX = byteArrayOf(2)
private val PARAMS_PREFIX = byteArrayOf(7)

/**
 * Contract handles transaction processing logic
 */
class Contract(
    val config: Config,
    val fsmConfig: PluginFSMConfig?,
    val plugin: PluginClient,
    val fsmId: Long
) {
    /**
     * Genesis implements logic to import/export state at height 0
     */
    fun genesis(request: PluginGenesisRequest): PluginGenesisResponse {
        logger.debug { "Processing genesis request" }
        return PluginGenesisResponse.getDefaultInstance()
    }

    /**
     * BeginBlock is executed at the start of applying a block
     */
    fun beginBlock(request: PluginBeginRequest): PluginBeginResponse {
        logger.debug { "Processing begin block" }
        return PluginBeginResponse.getDefaultInstance()
    }

    /**
     * CheckTx validates a transaction statelessly
     */
    fun checkTx(request: PluginCheckRequest): PluginCheckResponse {
        logger.debug { "Processing check tx" }

        // Validate fee by reading fee params
        val feeParamsKey = keyForFeeParams()
        val readRequest = PluginStateReadRequest.newBuilder()
            .addKeys(PluginKeyRead.newBuilder()
                .setQueryId(Random.nextLong())
                .setKey(ByteString.copyFrom(feeParamsKey))
                .build())
            .build()

        val readResponse = plugin.stateRead(this, readRequest)

        if (readResponse.hasError() && readResponse.error.code != 0L) {
            return PluginCheckResponse.newBuilder()
                .setError(readResponse.error)
                .build()
        }

        // Parse fee params
        if (readResponse.resultsCount > 0 && readResponse.getResults(0).entriesCount > 0) {
            val feeParamsBytes = readResponse.getResults(0).getEntries(0).value.toByteArray()
            if (feeParamsBytes.isNotEmpty()) {
                val feeParams = FeeParams.parseFrom(feeParamsBytes)
                if (request.tx.fee < feeParams.sendFee) {
                    return PluginCheckResponse.newBuilder()
                        .setError(ErrTxFeeBelowStateLimit().toProto())
                        .build()
                }
            }
        }

        // Unpack the message
        val msg = fromAny(request.tx.msg)
            ?: return PluginCheckResponse.newBuilder()
                .setError(ErrInvalidMessageCast().toProto())
                .build()

        return when (msg) {
            is MessageSend -> checkMessageSend(msg)
            else -> PluginCheckResponse.newBuilder()
                .setError(ErrInvalidMessageCast().toProto())
                .build()
        }
    }

    /**
     * DeliverTx applies a transaction
     */
    fun deliverTx(request: PluginDeliverRequest): PluginDeliverResponse {
        logger.debug { "Processing deliver tx" }

        val msg = fromAny(request.tx.msg)
            ?: return PluginDeliverResponse.newBuilder()
                .setError(ErrInvalidMessageCast().toProto())
                .build()

        return when (msg) {
            is MessageSend -> deliverMessageSend(msg, request.tx.fee)
            else -> PluginDeliverResponse.newBuilder()
                .setError(ErrInvalidMessageCast().toProto())
                .build()
        }
    }

    /**
     * EndBlock is executed at the end of applying a block
     */
    fun endBlock(request: PluginEndRequest): PluginEndResponse {
        logger.debug { "Processing end block" }
        return PluginEndResponse.getDefaultInstance()
    }

    /**
     * CheckMessageSend validates a send message statelessly
     */
    private fun checkMessageSend(msg: MessageSend): PluginCheckResponse {
        // Check sender address (must be 20 bytes)
        if (msg.fromAddress.size() != 20) {
            return PluginCheckResponse.newBuilder()
                .setError(ErrInvalidAddress().toProto())
                .build()
        }

        // Check recipient address (must be 20 bytes)
        if (msg.toAddress.size() != 20) {
            return PluginCheckResponse.newBuilder()
                .setError(ErrInvalidAddress().toProto())
                .build()
        }

        // Check amount
        if (msg.amount == 0L) {
            return PluginCheckResponse.newBuilder()
                .setError(ErrInvalidAmount().toProto())
                .build()
        }

        return PluginCheckResponse.newBuilder()
            .setRecipient(msg.toAddress)
            .addAuthorizedSigners(msg.fromAddress)
            .build()
    }

    /**
     * DeliverMessageSend handles a send message
     */
    private fun deliverMessageSend(msg: MessageSend, fee: Long): PluginDeliverResponse {
        val fromKey = keyForAccount(msg.fromAddress.toByteArray())
        val toKey = keyForAccount(msg.toAddress.toByteArray())
        val feePoolKey = keyForFeePool(config.chainId)

        val fromQueryId = Random.nextLong()
        val toQueryId = Random.nextLong()
        val feeQueryId = Random.nextLong()

        // Read accounts and fee pool
        val readRequest = PluginStateReadRequest.newBuilder()
            .addKeys(PluginKeyRead.newBuilder().setQueryId(feeQueryId).setKey(ByteString.copyFrom(feePoolKey)).build())
            .addKeys(PluginKeyRead.newBuilder().setQueryId(fromQueryId).setKey(ByteString.copyFrom(fromKey)).build())
            .addKeys(PluginKeyRead.newBuilder().setQueryId(toQueryId).setKey(ByteString.copyFrom(toKey)).build())
            .build()

        val readResponse = plugin.stateRead(this, readRequest)

        if (readResponse.hasError() && readResponse.error.code != 0L) {
            return PluginDeliverResponse.newBuilder()
                .setError(readResponse.error)
                .build()
        }

        // Parse results
        var fromBytes: ByteArray = byteArrayOf()
        var toBytes: ByteArray = byteArrayOf()
        var feePoolBytes: ByteArray = byteArrayOf()

        for (result in readResponse.resultsList) {
            when (result.queryId) {
                fromQueryId -> if (result.entriesCount > 0) fromBytes = result.getEntries(0).value.toByteArray()
                toQueryId -> if (result.entriesCount > 0) toBytes = result.getEntries(0).value.toByteArray()
                feeQueryId -> if (result.entriesCount > 0) feePoolBytes = result.getEntries(0).value.toByteArray()
            }
        }

        // Parse accounts
        val from = if (fromBytes.isNotEmpty()) Account.parseFrom(fromBytes) else Account.getDefaultInstance()
        var to = if (toBytes.isNotEmpty()) Account.parseFrom(toBytes) else Account.getDefaultInstance()
        val feePool = if (feePoolBytes.isNotEmpty()) Pool.parseFrom(feePoolBytes) else Pool.getDefaultInstance()

        val amountToDeduct = msg.amount + fee

        // Check sufficient funds
        if (from.amount < amountToDeduct) {
            return PluginDeliverResponse.newBuilder()
                .setError(ErrInsufficientFunds().toProto())
                .build()
        }

        // For self-transfer, use same account
        if (fromKey.contentEquals(toKey)) {
            to = from
        }

        // Update balances
        val newFrom = from.toBuilder().setAmount(from.amount - amountToDeduct).build()
        val newTo = to.toBuilder().setAmount(to.amount + msg.amount).build()
        val newFeePool = feePool.toBuilder().setAmount(feePool.amount + fee).build()

        // Write state
        val writeRequest = if (newFrom.amount == 0L) {
            // Delete drained account
            PluginStateWriteRequest.newBuilder()
                .addSets(PluginSetOp.newBuilder().setKey(ByteString.copyFrom(feePoolKey)).setValue(ByteString.copyFrom(newFeePool.toByteArray())).build())
                .addSets(PluginSetOp.newBuilder().setKey(ByteString.copyFrom(toKey)).setValue(ByteString.copyFrom(newTo.toByteArray())).build())
                .addDeletes(PluginDeleteOp.newBuilder().setKey(ByteString.copyFrom(fromKey)).build())
                .build()
        } else {
            PluginStateWriteRequest.newBuilder()
                .addSets(PluginSetOp.newBuilder().setKey(ByteString.copyFrom(feePoolKey)).setValue(ByteString.copyFrom(newFeePool.toByteArray())).build())
                .addSets(PluginSetOp.newBuilder().setKey(ByteString.copyFrom(toKey)).setValue(ByteString.copyFrom(newTo.toByteArray())).build())
                .addSets(PluginSetOp.newBuilder().setKey(ByteString.copyFrom(fromKey)).setValue(ByteString.copyFrom(newFrom.toByteArray())).build())
                .build()
        }

        val writeResponse = plugin.stateWrite(this, writeRequest)

        return if (writeResponse.hasError() && writeResponse.error.code != 0L) {
            PluginDeliverResponse.newBuilder().setError(writeResponse.error).build()
        } else {
            PluginDeliverResponse.getDefaultInstance()
        }
    }
}

/**
 * Convert PluginError to protobuf PluginError
 */
fun com.canopy.plugin.PluginError.toProto(): types.Plugin.PluginError =
    types.Plugin.PluginError.newBuilder()
        .setCode(this.code.toLong())
        .setModule(this.module)
        .setMsg(this.msg)
        .build()

/**
 * Unpack Any to specific message type
 */
fun fromAny(any: Any?): com.google.protobuf.Message? {
    if (any == null) return null
    return try {
        when {
            any.typeUrl.endsWith("MessageSend") -> MessageSend.parseFrom(any.value)
            else -> null
        }
    } catch (e: Exception) {
        logger.error(e) { "Failed to unpack Any message" }
        null
    }
}

/**
 * Extension function to convert ByteArray to hex string for logging
 */
@OptIn(ExperimentalStdlibApi::class)
private fun ByteArray.toHexString(): String = this.toHexString(HexFormat.Default)

/**
 * Key generation functions matching Go implementation
 */
fun keyForAccount(addr: ByteArray): ByteArray = joinLenPrefix(ACCOUNT_PREFIX, addr)

fun keyForFeeParams(): ByteArray = joinLenPrefix(PARAMS_PREFIX, "/f/".toByteArray())

fun keyForFeePool(chainId: Long): ByteArray = joinLenPrefix(POOL_PREFIX, formatUint64(chainId))

/**
 * Format uint64 as big-endian bytes
 */
private fun formatUint64(value: Long): ByteArray {
    val buffer = ByteBuffer.allocate(8).order(ByteOrder.BIG_ENDIAN)
    buffer.putLong(value)
    return buffer.array()
}

/**
 * Join byte arrays with length prefixes (matching Go's JoinLenPrefix)
 */
private fun joinLenPrefix(vararg items: ByteArray): ByteArray {
    val totalLen = items.sumOf { if (it.isNotEmpty()) 1 + it.size else 0 }
    val result = ByteArray(totalLen)
    var pos = 0
    for (item in items) {
        if (item.isEmpty()) continue
        result[pos++] = item.size.toByte()
        System.arraycopy(item, 0, result, pos, item.size)
        pos += item.size
    }
    return result
}
